riscv/src/compiler.rs

```diff
@@ -1416,44 +1297,29 @@ fn process_instruction(instr: &str, args: &[Argument], coprocessors: &CoProcesso
             vec![format!("x1 <== jump_dyn({rs});")]
         }
         "call" | "tail" => {
-            // Depending on what symbol is called, the call is replaced by a
-            // powdr-asm call, or a call to a coprocessor if a special function
-            // has been recognized.
             assert_eq!(args.len(), 1);
             let label = &args[0];
-            let replacement = match label {
-                Argument::Expression(Expression::Symbol(l)) => {
-                    try_coprocessor_substitution(l, coprocessors)
-                }
-                _ => None,
-            };
-            match (replacement, instr) {
-                (None, instr) => {
-                    let arg = argument_to_escaped_symbol(label);
-                    let dest = if instr == "tail" { "tmp1" } else { "x1" };
-                    vec![format!("{dest} <== jump({arg});")]
-                }
-                // Both "call" and "tail" are pseudo-instructions that are
-                // supposed to use x6 to calculate the high bits of the
-                // destination address. Our implementation does not touch x6,
-                // but no sane program would rely on this behavior, so we are
-                // probably fine.
-                (Some(replacement), "call") => vec![replacement],
-                (Some(replacement), "tail") => {
-                    vec![replacement, "tmp1 <== jump_dyn(x1);".to_string()]
-                }
-                (Some(_), _) => unreachable!(),
-            }
+            let arg = argument_to_escaped_symbol(label);
+            let dest = if instr == "tail" { "tmp1" } else { "x1" };
+            vec![format!("{dest} <== jump({arg});")]
         }
         "ecall" => {
             assert!(args.is_empty());
-            vec!["x10 <=X= ${ std::prover::Query::Input(std::convert::int(std::prover::eval(x10))) };".to_string()]
+            vec![
+                // push ra/x1 into stack
+                "x2 <=X= wrap(x2 - 4);".to_string(),
+                "mstore x2, x1;".to_string(),
+                // jump to to handler
+                "x1 <== jump(__ecall_handler);".to_string(),
+                // pop ra/x1
+                "x1, tmp1 <== mload(x2);".to_string(), // restore ra
+                "x2 <=X= wrap(x2 + 4);".to_string(),
```

Comment:
Could have a helper for this

Reply:
this whole block of powdr asm

Reply:
idk, looks ok to me

Reply:
lint isn't complaining so I guess it's fine

Reply:
wait, ignore my previous comment, I confused the topic

Reply:
here I just meant this block of asm could be useful for other cases where we might need the same thing. But fine with waiting for that to happen

Reply:
i refactored the push/pop to stack to helper methods